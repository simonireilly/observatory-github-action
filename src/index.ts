import * as core from '@actions/core'
import * as exec from '@actions/exec'

type JSONReport = {
  [key: string]: { [key: string]: string }
}

const webHost = (): string => {
  return core.getInput('web_host') || 'github.com'
}

export async function run(): Promise<string> {
  let sanitizedHostName = webHost()

  try {
    sanitizedHostName = new URL(webHost()).host
  } catch (e) {
    core.warning('This is not a valid URL, trying as given string')
    core.error(e)
  }

  core.debug(`Running on website: ${sanitizedHostName}`)

  const { result, error } = await runObservatory(sanitizedHostName)

  if (error) core.debug(error)

  core.debug(result)

  const markdown = jsonReportToMarkdown(result, sanitizedHostName)

  core.setOutput('observatory-report', markdown)
  return markdown
}

export async function runObservatory(
  sanitizedHostName: string
): Promise<{
  result: string
  error?: string
}> {
  let result = ''
  let error = ''

  const options = {
    listeners: {}
  }

  options.listeners = {
    stdout: (data: Buffer) => {
      result += data.toString()
    },
    stderr: (data: Buffer) => {
      error += data.toString()
    }
  }

  try {
    await exec.exec(
      'npx',
      ['observatory-cli', sanitizedHostName, '--format=json', '--attempts=30'],
      options
    )
  } catch (e) {
    core.setFailed(e.message)
  }

  core.info(result)
  core.error(error)

  return { result, error }
}

export function jsonReportToMarkdown(
  jsonReport: JSONReport | string,
  sanitizedHostName: string
): string {
  let result

  if (typeof jsonReport === 'string') {
    const jsonStructure = jsonReport.slice(jsonReport.indexOf('{'))
    if (jsonStructure.length > 0) {
      result = JSON.parse(jsonStructure) as JSONReport
    } else {
      core.setFailed('Result is empty')
      return ''
    }
  } else {
    result = jsonReport
  }

  const resultRows: string[] = []
  let score = 100

  // Get the keys
  for (const key in result) {
    const { score_modifier = '0', pass, score_description } = result[key]
    const success = Boolean(pass)

    score += parseInt(score_modifier)
    const icon = (showSuccessIcon: boolean): string =>
      showSuccessIcon ? ':green_circle:' : ':red_circle:'
    const message = `${icon(
      success
    )} | ${score_modifier} | ${score_description}`
    resultRows.push(message)
  }

  return `
## Observatory Results [${sanitizedHostName}](https://${sanitizedHostName}): _${score} of 100_

See the full report: https://observatory.mozilla.org/analyze/${sanitizedHostName}

### Highlights

Passed | Score | Description
--- | --- | ---
${resultRows.join('\n')}
  `
}
