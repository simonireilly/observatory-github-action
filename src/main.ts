import * as core from '@actions/core';
import { run } from './index';

try {
  run();
} catch (error) {
  if (error instanceof Error) {
    core.setFailed(error.message);
  } else {
    core.setFailed('An unknown error occurred');
  }
}
