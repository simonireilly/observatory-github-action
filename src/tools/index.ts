// Handles configuration, and calling out to the runner
import { runObservatory } from './observatory-runner';
import * as core from '@actions/core';
import { jsonReportToMarkdown } from './report-generator';

const webHost = (): string => {
  return core.getInput('web_host') || 'github.com';
};

export async function run(): Promise<string> {
  let sanitizedHostName = webHost();

  try {
    sanitizedHostName = new URL(webHost()).host;
  } catch (e: unknown) {
    core.warning('This is not a valid URL, trying as given string');
    if (e instanceof Error) {
      core.error(e);
    } else {
      core.setFailed('An unknown error occurred');
    }
  }

  core.debug(`Running on website: ${sanitizedHostName}`);

  const { result, error } = await runObservatory(sanitizedHostName);

  if (error) core.debug(error);

  core.debug(result);

  const markdown = jsonReportToMarkdown(result, sanitizedHostName);

  core.setOutput('observatory-report', markdown);
  return markdown;
}
