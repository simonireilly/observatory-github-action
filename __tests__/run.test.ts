import { run, runObservatory, jsonReportToMarkdown } from '../src/index'

jest.setTimeout(10000)

describe('run', () => {
  it('returns a full comment from the observatory client', async () => {
    const markdownString = await run()

    expect(markdownString).toMatchSnapshot()
  })
})

describe('jsonReportToMarkdown', () => {
  it('Creates a table for the results', () => {
    const jsonReport = require('./fixtures/json-observatory-report.json')

    const markdown = jsonReportToMarkdown(jsonReport)

    expect(markdown).toMatchSnapshot()
  })
})
