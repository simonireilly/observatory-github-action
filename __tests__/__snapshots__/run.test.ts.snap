// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`jsonReportToMarkdown Creates a table for the results 1`] = `
"
### Observatory Results

0 of 100

#### Breakdown

Passed | Score | Description
--- | --- | ---
:red_circle: | -25 | Content Security Policy (CSP) header not implemented
:green_circle: | 0 | Contribute.json isn't required on websites that don't belong to Mozilla
:green_circle: | 0 | No cookies detected
:green_circle: | 0 | Content is not visible via cross-origin resource sharing (CORS) files or headers
:green_circle: | 0 | HTTP Public Key Pinning (HPKP) header not implemented
:red_circle: | -20 | Does not redirect to an HTTPS site
:green_circle: | 0 | Referrer-Policy header not implemented
:red_circle: | -20 | HTTP Strict Transport Security (HSTS) header not implemented
:green_circle: | 0 | Subresource Integrity (SRI) is not needed since site contains no script tags
:red_circle: | -5 | X-Content-Type-Options header not implemented
:red_circle: | -20 | X-Frame-Options (XFO) header not implemented
:red_circle: | -10 | X-XSS-Protection header not implemented
  "
`;

exports[`run returns a full comment from the observatory client 1`] = `
"
### Observatory Results

125 of 100

#### Breakdown

Passed | Score | Description
--- | --- | ---
:green_circle: | 0 | Content Security Policy (CSP) implemented with unsafe sources inside style-src. This includes 'unsafe-inline', data: or overly broad sources such as https:.
:green_circle: | 0 | Contribute.json isn't required on websites that don't belong to Mozilla
:green_circle: | 5 | All cookies use the Secure flag, session cookies use the HttpOnly flag, and cross-origin restrictions are in place via the SameSite flag
:green_circle: | 0 | Content is not visible via cross-origin resource sharing (CORS) files or headers
:green_circle: | 0 | HTTP Public Key Pinning (HPKP) header not implemented
:green_circle: | 0 | All hosts redirected to are in the HTTP Strict Transport Security (HSTS) preload list
:green_circle: | 5 | Referrer-Policy header set to \\"no-referrer\\", \\"same-origin\\", \\"strict-origin\\" or \\"strict-origin-when-cross-origin\\"
:green_circle: | 5 | Preloaded via the HTTP Strict Transport Security (HSTS) preloading process
:green_circle: | 5 | Subresource Integrity (SRI) is implemented and all scripts are loaded securely
:green_circle: | 0 | X-Content-Type-Options header set to \\"nosniff\\"
:green_circle: | 5 | X-Frame-Options (XFO) implemented via the CSP frame-ancestors directive
:green_circle: | 0 | X-XSS-Protection header set to \\"1; mode=block\\"
  "
`;

exports[`runObservatory returns a formatted JSON payload 1`] = `
"{
  \\"content-security-policy\\": {
    \\"expectation\\": \\"csp-implemented-with-no-unsafe\\",
    \\"name\\": \\"content-security-policy\\",
    \\"output\\": {
      \\"data\\": {
        \\"base-uri\\": [
          \\"'self'\\"
        ],
        \\"block-all-mixed-content\\": [
          \\"'none'\\"
        ],
        \\"connect-src\\": [
          \\"github-production-user-asset-6210df.s3.amazonaws.com\\",
          \\"api.github.com\\",
          \\"github-cloud.s3.amazonaws.com\\",
          \\"logx.optimizely.com/v1/events\\",
          \\"cdn.optimizely.com\\",
          \\"wss://alive.github.com\\",
          \\"github-production-repository-file-5c1aeb.s3.amazonaws.com\\",
          \\"'self'\\",
          \\"www.githubstatus.com\\",
          \\"uploads.github.com\\",
          \\"collector.githubapp.com\\",
          \\"github-production-upload-manifest-file-7fdce7.s3.amazonaws.com\\"
        ],
        \\"default-src\\": [
          \\"'none'\\"
        ],
        \\"font-src\\": [
          \\"github.githubassets.com\\"
        ],
        \\"form-action\\": [
          \\"gist.github.com\\",
          \\"github.com\\",
          \\"'self'\\"
        ],
        \\"frame-ancestors\\": [
          \\"'none'\\"
        ],
        \\"frame-src\\": [
          \\"render.githubusercontent.com\\"
        ],
        \\"img-src\\": [
          \\"customer-stories-feed.github.com\\",
          \\"github-cloud.s3.amazonaws.com\\",
          \\"*.githubusercontent.com\\",
          \\"github.githubassets.com\\",
          \\"spotlights-feed.github.com\\",
          \\"data:\\",
          \\"identicons.github.com\\",
          \\"'self'\\",
          \\"collector.githubapp.com\\"
        ],
        \\"manifest-src\\": [
          \\"'self'\\"
        ],
        \\"media-src\\": [
          \\"github.githubassets.com\\"
        ],
        \\"script-src\\": [
          \\"github.githubassets.com\\"
        ],
        \\"style-src\\": [
          \\"github.githubassets.com\\",
          \\"'unsafe-inline'\\"
        ],
        \\"worker-src\\": [
          \\"github.com/socket-worker-5029ae85.js\\",
          \\"gist.github.com/socket-worker-5029ae85.js\\"
        ]
      },
      \\"http\\": true,
      \\"meta\\": false,
      \\"policy\\": {
        \\"antiClickjacking\\": true,
        \\"defaultNone\\": true,
        \\"insecureBaseUri\\": false,
        \\"insecureFormAction\\": false,
        \\"insecureSchemeActive\\": false,
        \\"insecureSchemePassive\\": false,
        \\"strictDynamic\\": false,
        \\"unsafeEval\\": false,
        \\"unsafeInline\\": false,
        \\"unsafeInlineStyle\\": true,
        \\"unsafeObjects\\": false
      }
    },
    \\"pass\\": true,
    \\"result\\": \\"csp-implemented-with-unsafe-inline-in-style-src-only\\",
    \\"score_description\\": \\"Content Security Policy (CSP) implemented with unsafe sources inside style-src. This includes 'unsafe-inline', data: or overly broad sources such as https:.\\",
    \\"score_modifier\\": 0
  },
  \\"contribute\\": {
    \\"expectation\\": \\"contribute-json-only-required-on-mozilla-properties\\",
    \\"name\\": \\"contribute\\",
    \\"output\\": {
      \\"data\\": null
    },
    \\"pass\\": true,
    \\"result\\": \\"contribute-json-only-required-on-mozilla-properties\\",
    \\"score_description\\": \\"Contribute.json isn't required on websites that don't belong to Mozilla\\",
    \\"score_modifier\\": 0
  },
  \\"cookies\\": {
    \\"expectation\\": \\"cookies-secure-with-httponly-sessions\\",
    \\"name\\": \\"cookies\\",
    \\"output\\": {
      \\"data\\": {
        \\"_gh_sess\\": {
          \\"domain\\": \\"github.com\\",
          \\"expires\\": null,
          \\"httponly\\": true,
          \\"max-age\\": null,
          \\"path\\": \\"/\\",
          \\"port\\": null,
          \\"samesite\\": \\"Lax\\",
          \\"secure\\": true
        },
        \\"_octo\\": {
          \\"domain\\": \\".github.com\\",
          \\"expires\\": 1640266828,
          \\"httponly\\": false,
          \\"max-age\\": null,
          \\"path\\": \\"/\\",
          \\"port\\": null,
          \\"samesite\\": \\"Lax\\",
          \\"secure\\": true
        },
        \\"logged_in\\": {
          \\"domain\\": \\".github.com\\",
          \\"expires\\": 1640266828,
          \\"httponly\\": true,
          \\"max-age\\": null,
          \\"path\\": \\"/\\",
          \\"port\\": null,
          \\"samesite\\": \\"Lax\\",
          \\"secure\\": true
        }
      },
      \\"sameSite\\": true
    },
    \\"pass\\": true,
    \\"result\\": \\"cookies-secure-with-httponly-sessions-and-samesite\\",
    \\"score_description\\": \\"All cookies use the Secure flag, session cookies use the HttpOnly flag, and cross-origin restrictions are in place via the SameSite flag\\",
    \\"score_modifier\\": 5
  },
  \\"cross-origin-resource-sharing\\": {
    \\"expectation\\": \\"cross-origin-resource-sharing-not-implemented\\",
    \\"name\\": \\"cross-origin-resource-sharing\\",
    \\"output\\": {
      \\"data\\": {
        \\"acao\\": null,
        \\"clientaccesspolicy\\": null,
        \\"crossdomain\\": null
      }
    },
    \\"pass\\": true,
    \\"result\\": \\"cross-origin-resource-sharing-not-implemented\\",
    \\"score_description\\": \\"Content is not visible via cross-origin resource sharing (CORS) files or headers\\",
    \\"score_modifier\\": 0
  },
  \\"public-key-pinning\\": {
    \\"expectation\\": \\"hpkp-not-implemented\\",
    \\"name\\": \\"public-key-pinning\\",
    \\"output\\": {
      \\"data\\": null,
      \\"includeSubDomains\\": false,
      \\"max-age\\": null,
      \\"numPins\\": null,
      \\"preloaded\\": false
    },
    \\"pass\\": true,
    \\"result\\": \\"hpkp-not-implemented\\",
    \\"score_description\\": \\"HTTP Public Key Pinning (HPKP) header not implemented\\",
    \\"score_modifier\\": 0
  },
  \\"redirection\\": {
    \\"expectation\\": \\"redirection-to-https\\",
    \\"name\\": \\"redirection\\",
    \\"output\\": {
      \\"destination\\": \\"https://github.com/\\",
      \\"redirects\\": true,
      \\"route\\": [
        \\"http://github.com/\\",
        \\"https://github.com/\\"
      ],
      \\"status_code\\": 200
    },
    \\"pass\\": true,
    \\"result\\": \\"redirection-all-redirects-preloaded\\",
    \\"score_description\\": \\"All hosts redirected to are in the HTTP Strict Transport Security (HSTS) preload list\\",
    \\"score_modifier\\": 0
  },
  \\"referrer-policy\\": {
    \\"expectation\\": \\"referrer-policy-private\\",
    \\"name\\": \\"referrer-policy\\",
    \\"output\\": {
      \\"data\\": \\"origin-when-cross-origin, strict-origin-when-cross-origin\\",
      \\"http\\": true,
      \\"meta\\": false
    },
    \\"pass\\": true,
    \\"result\\": \\"referrer-policy-private\\",
    \\"score_description\\": \\"Referrer-Policy header set to \\\\\\"no-referrer\\\\\\", \\\\\\"same-origin\\\\\\", \\\\\\"strict-origin\\\\\\" or \\\\\\"strict-origin-when-cross-origin\\\\\\"\\",
    \\"score_modifier\\": 5
  },
  \\"strict-transport-security\\": {
    \\"expectation\\": \\"hsts-implemented-max-age-at-least-six-months\\",
    \\"name\\": \\"strict-transport-security\\",
    \\"output\\": {
      \\"data\\": \\"max-age=31536000; includeSubdomains; preload\\",
      \\"includeSubDomains\\": true,
      \\"max-age\\": 31536000,
      \\"preload\\": true,
      \\"preloaded\\": true
    },
    \\"pass\\": true,
    \\"result\\": \\"hsts-preloaded\\",
    \\"score_description\\": \\"Preloaded via the HTTP Strict Transport Security (HSTS) preloading process\\",
    \\"score_modifier\\": 5
  },
  \\"subresource-integrity\\": {
    \\"expectation\\": \\"sri-implemented-and-external-scripts-loaded-securely\\",
    \\"name\\": \\"subresource-integrity\\",
    \\"output\\": {
      \\"data\\": {
        \\"https://github.githubassets.com/assets/behaviors-824642a4.js\\": {
          \\"crossorigin\\": \\"anonymous\\",
          \\"integrity\\": \\"sha512-gkZCpFyYrX7ycwoeUW7Wyk7sg8OBhVVIA5gFuJfzX3jZcu1jiTPWWM749XEjX9pILro9F5+l+WUfz/MUpAwI0w==\\"
        },
        \\"https://github.githubassets.com/assets/chunk-frameworks-6dfde4f1.js\\": {
          \\"crossorigin\\": \\"anonymous\\",
          \\"integrity\\": \\"sha512-bf3k8Y60Hljk6JoJikJwUktgpWsNM3NEnNgmlScAD6u57WAUEgCSZh5BsU5ls8YxsL8cZUaKGHC5fsf143pfaw==\\"
        },
        \\"https://github.githubassets.com/assets/chunk-vendor-4d97ead9.js\\": {
          \\"crossorigin\\": \\"anonymous\\",
          \\"integrity\\": \\"sha512-TZfq2T+XE31Mcp17bGlABjjeQ/69IH2f/sBvS31c4Jua0e1O/NE+aVbQtItMM2GFlmeEm13p/zyyWGIx10jVKQ==\\"
        },
        \\"https://github.githubassets.com/assets/environment-f0adafbf.js\\": {
          \\"crossorigin\\": \\"anonymous\\",
          \\"integrity\\": \\"sha512-8K2vvwbW+6H27Nad5ydg8PA2/aMD/LKq+EiK9s0U0hhVZxCI2tWBsYk9beAtisRw2j+Or5k2/F+6dk02nmj/PA==\\"
        },
        \\"https://github.githubassets.com/assets/features-5d0ec104.js\\": {
          \\"crossorigin\\": \\"anonymous\\",
          \\"integrity\\": \\"sha512-XQ7BBIQKF+bAgaheKl7KNH34ik/T4ndXTF37yg4H4ojAZ02CXqxaG+kjwbfMiXRbzZNw8x8YkRIYeQYr5faYbA==\\"
        },
        \\"https://github.githubassets.com/assets/marketing-28b1ca73.js\\": {
          \\"crossorigin\\": \\"anonymous\\",
          \\"integrity\\": \\"sha512-KLHKc0eJEbmI24nM12O7c2ELsFsSxlYUP5gcuww2HQg/dskU6B9XzeFXzwPfgVrvtZLhXnq61JkD7IHlC+RYEg==\\"
        },
        \\"https://github.githubassets.com/assets/site-874e15eb.js\\": {
          \\"crossorigin\\": \\"anonymous\\",
          \\"integrity\\": \\"sha512-h04V63L89puz9DQJ2prtxxDmBXOh5NX4oqlgJtGYIFpUggs8sttjjxmbgTU4872MmCfeWsoCSXQPFn8lxjtGQA==\\"
        },
        \\"https://github.githubassets.com/assets/unsupported-a85b1284.js\\": {
          \\"crossorigin\\": \\"anonymous\\",
          \\"integrity\\": \\"sha512-qFsShJX3EkHdcQq11CLfRk444sM6/0OBXB8eTN3FZl70HSy6jUPI2M9H6/wNWDwOR+LLU/JE55Y2kl1CK1QioQ==\\"
        },
        \\"https://github.githubassets.com/assets/webgl-globe-89c68dda.js\\": {
          \\"crossorigin\\": \\"anonymous\\",
          \\"integrity\\": \\"sha512-icaN2toWhKufebiGwejYDV1K03YNiUqCf7wry+/ZYHlNwPpqo+F5w1LlNUTdCL4+ZA2Xx0mjNcOFaiLxiPd1pA==\\"
        }
      }
    },
    \\"pass\\": true,
    \\"result\\": \\"sri-implemented-and-external-scripts-loaded-securely\\",
    \\"score_description\\": \\"Subresource Integrity (SRI) is implemented and all scripts are loaded securely\\",
    \\"score_modifier\\": 5
  },
  \\"x-content-type-options\\": {
    \\"expectation\\": \\"x-content-type-options-nosniff\\",
    \\"name\\": \\"x-content-type-options\\",
    \\"output\\": {
      \\"data\\": \\"nosniff\\"
    },
    \\"pass\\": true,
    \\"result\\": \\"x-content-type-options-nosniff\\",
    \\"score_description\\": \\"X-Content-Type-Options header set to \\\\\\"nosniff\\\\\\"\\",
    \\"score_modifier\\": 0
  },
  \\"x-frame-options\\": {
    \\"expectation\\": \\"x-frame-options-sameorigin-or-deny\\",
    \\"name\\": \\"x-frame-options\\",
    \\"output\\": {
      \\"data\\": \\"deny\\"
    },
    \\"pass\\": true,
    \\"result\\": \\"x-frame-options-implemented-via-csp\\",
    \\"score_description\\": \\"X-Frame-Options (XFO) implemented via the CSP frame-ancestors directive\\",
    \\"score_modifier\\": 5
  },
  \\"x-xss-protection\\": {
    \\"expectation\\": \\"x-xss-protection-1-mode-block\\",
    \\"name\\": \\"x-xss-protection\\",
    \\"output\\": {
      \\"data\\": \\"1; mode=block\\"
    },
    \\"pass\\": true,
    \\"result\\": \\"x-xss-protection-enabled-mode-block\\",
    \\"score_description\\": \\"X-XSS-Protection header set to \\\\\\"1; mode=block\\\\\\"\\",
    \\"score_modifier\\": 0
  }
}
"
`;
